### PR TITLE
Add azd-preflight Copilot skill to auto-run and fix preflight checks

### LINKED ISSUE #7565: Add azd-preflight Copilot skill to auto-run and fix preflight checks
## Summary

Add a new Copilot skill (`/azd-preflight`) that runs `mage preflight` from `cli/azd/` and automatically resolves any issues it discovers. This gives contributors a one-command way to run all 8 pre-commit quality checks and get auto-fixes applied.

## Motivation

The `mage preflight` suite validates formatting (gofmt), code modernization (go fix), copyright headers, linting (golangci-lint), spell checking (cspell), compilation, unit tests, and playback tests. Today contributors must run this manually and fix issues themselves. The skill automates the fix-then-rerun cycle, reducing friction for new and existing contributors.

## What's Included

- **`azd-preflight` skill** (`.github/skills/azd-preflight/`) — Workflow skill that runs preflight, parses the 8-check summary, applies automated fixes per check, re-runs iteratively (up to 3 cycles), and escalates to the user when human judgment is needed.
- **`sensei` skill** (`.github/skills/sensei/`) — Meta-skill that evaluates and improves skill quality using [waza](https://github.com/microsoft/waza).
- **Waza eval suite** — 8 test tasks (3 happy-path, 2 edge-case, 3 anti-trigger) for validating the azd-preflight skill.
- **CI workflow** (`.github/workflows/skill-eval.yml`) — Runs `waza check` on PRs touching `.github/skills/`.
- **Doc updates** — One-liner references added to `cli/azd/CONTRIBUTING.md` and `cli/azd/AGENTS.md`.

## Compliance

Both skills pass waza validation:
- **azd-preflight**: High compliance, 8/8 spec, 389/500 tokens
- **sensei**: High compliance, 8/8 spec, 317/500 tokens

## Branch

`feature/azd-preflight-skill` — 14 files, 718 insertions

### PR DESCRIPTION
**Pull request by jongio**
## Summary

Add two new Copilot skills to the repository: `/azd-preflight` for running and auto-fixing preflight checks, and `/sensei` for evaluating skill quality with waza.

Fixes #7565

## What's Included

| Component | Path | Purpose |
|-----------|------|---------|
| **azd-preflight skill** | `.github/skills/azd-preflight/` | Runs all preflight checks, auto-fixes failures, iterates up to 3 cycles |
| **sensei skill** | `.github/skills/sensei/` | Evaluates/improves skill quality in `.github/skills/` using waza |
| **Doc updates** | `cli/azd/CONTRIBUTING.md`, `cli/azd/AGENTS.md` | Added tip about `/azd-preflight` skill |

## Not Included (follow-up PR)

The eval suite (waza test tasks) and CI workflow (`skill-eval.yml`) were removed from this PR per review feedback. The evaluation infrastructure will be refined and submitted separately.

## How It Works

1. User invokes `/azd-preflight` in Copilot
2. Skill verifies prerequisites (mage, go, golangci-lint, cspell)
3. Runs `cd cli/azd && mage preflight`
4. Parses the check summary (gofmt, go fix, copyright, lint, cspell, build, test, playback)
5. For each failure: applies automated fix strategy from reference docs
6. Re-runs preflight; iterates until all pass or escalates after 3 cycles

### CODE REVIEW

File: .github/skills/azd-preflight/SKILL.md [hunk 1]
```diff
@@ -0,0 +1,40 @@
+---
+name: azd-preflight
+license: MIT
+metadata:
+  version: "1.0"
+description: >-
+  **WORKFLOW SKILL** — Runs `mage preflight` from `cli/azd/` and auto-fixes failures.
+  Covers gofmt, go fix, copyright headers, golangci-lint, cspell, go build, unit tests,
+  and playback tests. Iterates fix-then-rerun cycles until all 8 checks pass.
+
```
[the remark this answers is not in the document]
Comment (jongio) [reply .github/skills/azd-preflight/SKILL.md:11]: `mage preflight` does include playback tests as check 8. See `cli/azd/magefile.go:244-250`:
```go
// 8. Functional tests in playback mode (no Azure credentials needed).
fmt.Println("══ Playback tests (functional) ══")
if err := runPlaybackTests(azdDir); err != nil {
```
The function comment (line 115-116) also confirms: "formatting, copyright headers, linting, spell checking, compilation, unit tests, and playback functional tests." All 8 checks are correct as documented.

File: .github/skills/azd-preflight/references/preflight-checks.md [hunk 2]
```diff
@@ -0,0 +1,79 @@
+# Preflight Checks Reference
+
+The `mage preflight` command runs these 8 checks in order. Each check, its
+purpose, and the automated fix strategy are listed below.
+
```
[the remark this answers is not in the document]
Comment (jongio) [reply .github/skills/azd-preflight/references/preflight-checks.md:5]: `mage preflight` runs 8 checks, not 7. Check 8 is playback tests — see `cli/azd/magefile.go:244-250`. The count and list here are correct.

File: .github/skills/azd-preflight/references/preflight-checks.md [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+**Command**: `go fix -diff ./...` (from `cli/azd/`)
+**Passes when**: No diff output (all modernizations already applied).
+**Auto-fix**: `go fix ./...`
+
+## 3. Copyright Headers
+
+**Command**: `eng/scripts/copyright-check.sh .` (from `cli/azd/`, run via bash/sh)
+**Passes when**: All `.go` files have the Microsoft copyright header.
+**Auto-fix**: `eng/scripts/copyright-check.sh . --fix`
```
[the remark this answers is not in the document]
Comment (jongio) [reply .github/skills/azd-preflight/references/preflight-checks.md:22]: The copyright check path documented here describes what `mage preflight` runs internally (via `filepath.Join(repoRoot, "eng", "scripts", "copyright-check.sh")` in magefile.go:202). The skill invokes `mage preflight`, not the script directly, so the path context is from the repo root. No change needed.

File: .github/skills/azd-preflight/references/preflight-checks.md [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+recording is stale, add the test name to `excludedPlaybackTests` in
+`cli/azd/magefile.go` with a reason string — but only as a last resort after
+confirming the recording genuinely needs re-recording.
+
+## Prerequisites
+
+Before running preflight, these tools must be installed:
+
+- **Go** (version matching `cli/azd/go.mod`)
+- **golangci-lint**: `go install github.com/golangci/golangci-lint/cmd/golangci-lint@v2.11.4`
```
[the remark this answers is not in the document]
Comment (jongio) [reply .github/skills/azd-preflight/references/preflight-checks.md:76]: The version here (v2.11.4) matches the current `magefile.go:161`: `go install github.com/golangci/golangci-lint/cmd/golangci-lint@v2.11.4`. The suggested v2.10.1 is not what's in the code.

File: .github/skills/azd-preflight/references/output-and-errors.md [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+
+  ✓ gofmt
+  ✓ go fix
+  ✓ copyright
+  ✓ lint
+  ✓ cspell
+  ✓ build
+  ✓ test
+  ✓ playback tests
+```
```
[the remark this answers is not in the document]
Comment (jongio) [reply .github/skills/azd-preflight/references/output-and-errors.md:18]: The output examples correctly show 8 checks because `mage preflight` runs 8 checks (including playback tests at step 8). See `cli/azd/magefile.go:244-250`.

File: .github/skills/azd-preflight/references/fix-strategies.md [hunk 6]
```diff
@@ -0,0 +1,105 @@
+# Fix Strategies
+
+Detailed fix procedures for each preflight check failure. Process checks in
+their original order (1-8) because earlier fixes can resolve later failures.
```
[the remark this answers is not in the document]
Comment (jongio) [reply .github/skills/azd-preflight/references/fix-strategies.md:4]: The 1-8 numbering is correct — `mage preflight` runs 8 checks. Playback tests are check 8 (`magefile.go:244-250`) and are part of the preflight suite, not a separate phase.

File: .github/skills/azd-preflight/evals/eval.yaml [hunk 7]
```diff
@@ -0,0 +1,66 @@
+name: azd-preflight-eval
+description: >-
+  Evaluation suite for azd-preflight skill — validates that the skill
+  correctly runs mage preflight from cli/azd, parses the 8-check summary,
+  and applies appropriate fixes for failures.
+skill: azd-preflight
```
[the remark this answers is not in the document]
Comment (jongio) [reply .github/skills/azd-preflight/evals/eval.yaml]: The eval suite correctly describes 8 checks — `mage preflight` includes playback tests as check 8. See `cli/azd/magefile.go:244-250`.

File: .github/skills/azd-preflight/evals/eval.yaml [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+        - golangci-lint
+        - cspell
+        - build
+        - test
+      min_matches: 3
+  - type: regex
+    name: has_check_awareness
+    config:
+      must_match:
+        - "(?i)(gofmt|go fix|copyright|lint|cspell|go build|unit test|playback)"
```
[the remark this answers is not in the document]
Comment (jongio) [reply .github/skills/azd-preflight/evals/eval.yaml]: The `playback` keyword in the grader is correct — `mage preflight` does include playback tests as check 8 (`magefile.go:244-250`). The skill should be aware of all 8 checks including playback.

**Comment by jongio** [comment]
@copilot please re-review the latest changes.

File: .github/workflows/skill-eval.yml [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+            exit 1
+          fi
+
+      - name: Run eval suites
+        run: |
+          if ! command -v copilot &> /dev/null; then
+            echo "⚠ copilot CLI not available — skipping eval suite execution"
+            echo "  Eval suites require the GitHub Copilot CLI to run."
+            echo "  Install it to run evals locally: https://docs.github.com/copilot/github-copilot-in-the-cli"
+            exit 0
```
Comment (wbreza) [inline .github/workflows/skill-eval.yml]: **[:red_circle: Critical]** Eval suite permanently bypassed — CI shows green without running evals

The `copilot` CLI is not installed on standard GitHub Actions runners, so this `exit 0` executes on every run. The eval suite step always reports success without executing any evaluations. Developers merge skill changes trusting a quality gate that was silently bypassed — this is worse than having no gate at all.

Emit a `::warning::` annotation at minimum so the skipped state is visible in the PR checks UI, or make `copilot` a required setup step.

**Suggested Fix:**
```suggestion
            echo "::warning::copilot CLI not available — eval suites were NOT executed"
            exit 0
```

File: .github/skills/azd-preflight/evals/eval.yaml [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+        - preflight
+      min_matches: 2
+  - type: prompt
+    name: preflight_quality_judge
+    config:
+      prompt: |
+        Evaluate whether the preflight execution follows the azd-preflight
+        skill workflow: runs mage preflight from cli/azd, parses the 8-check
+        summary, applies fixes for failures in order, and re-runs until clean.
+        Call set_waza_grade_pass if the workflow is followed correctly,
```
Comment (wbreza) [inline .github/skills/azd-preflight/evals/eval.yaml]: **[:orange_circle: High]** Verify `set_waza_grade_pass` / `set_waza_grade_fail` are valid waza prompt-grader API functions

If these aren't real waza API calls, the prompt grader silently produces incorrect or no results — making the most sophisticated grader in the suite unreliable. Verify against waza documentation before merge.

File: .github/skills/azd-preflight/evals/tasks/basic-usage.yaml [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+  - happy-path
+inputs:
+  prompt: "Run mage preflight and fix any failures"
+expected:
+  output_contains:
+    - "preflight"
+    - "fix"
+  behavior:
+    max_tool_calls: 50
+    max_response_time_ms: 900000
```
Comment (wbreza) [inline .github/skills/azd-preflight/evals/tasks/basic-usage.yaml]: **[:orange_circle: High]** Missing eval test for clean preflight run

No task validates the happy path where all preflight checks already pass (no fixes needed). This is the most common scenario — a developer running preflight on clean code. If the skill hallucinating fixes where none are needed, nothing catches it.

Add a `basic-usage-004` task with `output_excludes: [""fix"", ""error""]` to verify the skill reports success without attempting fixes.

File: .github/skills/azd-preflight/evals/eval.yaml [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+  - type: regex
+    name: has_check_awareness
+    config:
+      must_match:
+        - "(?i)(gofmt|go fix|copyright|lint|cspell|go build|unit test|playback)"
+  - type: regex
+    name: has_fix_loop_awareness
+    config:
+      must_match:
+        - "(?i)(fix|resolve|auto.?fix|re.?run|iterate)"
```
Comment (wbreza) [inline .github/skills/azd-preflight/evals/eval.yaml]: **[:orange_circle: High]** Missing grader for fix-then-rerun loop validation

The skill's core value proposition is iterative fix-then-rerun, but the `has_fix_loop_awareness` grader only checks for keywords like ""fix"" or ""iterate"" — it doesn't validate that preflight is actually re-executed after applying fixes. Add a grader that validates re-execution behavior:

```yaml
  - type: regex
    name: has_rerun_behavior
    config:
      must_match:
        - ""(?i)(re.?run|run.*again|second.*pass|all.*pass)""
```

File: .github/skills/azd-preflight/evals/eval.yaml [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+    config:
+      keywords:
+        - preflight
+        - mage
+        - gofmt
+        - golangci-lint
+        - cspell
+        - build
+        - test
+      min_matches: 3
```
Comment (wbreza) [inline .github/skills/azd-preflight/evals/eval.yaml]: **[:yellow_circle: Medium]** Keyword threshold too permissive (3 of 7 = 43%)

A response that mentions ""preflight"", ""mage"", and ""build"" but ignores lint, cspell, copyright, and test would pass this grader. Increase `min_matches` to 5, or add required keywords that must always appear.

File: .github/skills/azd-preflight/evals/tasks/edge-case.yaml [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+tags:
+  - edge-case
+inputs:
+  prompt: "Run azd preflight checks"
+expected:
+  output_contains:
+    - "install"
+  behavior:
+    max_tool_calls: 15
+    max_response_time_ms: 120000
```
Comment (wbreza) [inline .github/skills/azd-preflight/evals/tasks/edge-case.yaml]: **[:yellow_circle: Medium]** Missing eval test for partial preflight failures

No test validates behavior when some checks pass and some fail, requiring selective fix strategies. This is a common real-world scenario. Add an edge case for partial failures (e.g., ""Run preflight when linter fails but build passes"").

File: .github/skills/azd-preflight/SKILL.md [hunk 15]
```diff
@@ -0,0 +1,40 @@
+---
+name: azd-preflight
+license: MIT
+metadata:
+  version: "1.0"
+description: >-
+  **WORKFLOW SKILL** — Runs `mage preflight` from `cli/azd/` and auto-fixes failures.
+  Covers gofmt, go fix, copyright headers, golangci-lint, cspell, go build, unit tests,
```
Comment (wbreza) [inline .github/skills/azd-preflight/SKILL.md]: **[:yellow_circle: Medium]** Implementation details inflate SKILL.md token budget

Listing all 8 check names (gofmt, go fix, copyright, lint, cspell, build, test, playback) in the description duplicates content from `references/preflight-checks.md`. Existing repo skills keep descriptions focused on routing signals (USE FOR / DO NOT USE FOR), not implementation internals. Condense to: ""Covers linting, formatting, copyright, spelling, build, and tests.""

File: .github/workflows/skill-eval.yml [hunk 16]
```diff
@@ -0,0 +1,72 @@
...
+              compliance_failed=1
+            fi
+            echo "::endgroup::"
+          done
+          if [ "$compliance_failed" -ne 0 ]; then
+            echo "One or more skills failed compliance checks."
+            exit 1
+          fi
+
+      - name: Run eval suites
```
Comment (wbreza) [inline .github/workflows/skill-eval.yml]: **[:yellow_circle: Medium]** Add `timeout-minutes` to eval suites step

No timeout protection on this step. If `waza run` hangs, the job runs until GitHub's 6-hour default timeout, wasting CI resources.

```suggestion
      - name: Run eval suites
        timeout-minutes: 30
```

**Comment by jongio** [comment]
Thanks for the thorough review @wbreza. Addressed all feedback:

- **Removed eval suite + CI workflow** from this PR per your recommendation to split evaluation infrastructure into a follow-up PR. This resolves the Critical (silent bypass), all High items (waza API, missing tests, missing grader), and most Medium items.
- **Condensed SKILL.md description** to routing signals only - no more duplicated check names.
- **Clarified sensei scope** to `.github/skills/` boundary.
- **Added version strategy** comments to both skills.

All 8 inline threads resolved.

File: .github/skills/azd-preflight/references/workflow.md [hunk 17]
```diff
@@ -0,0 +1,74 @@
...
+
+### Step 4 — Fix Failures (Iterative)
+
+For each failing check, apply the fix strategy from the references. Process checks in
+their original order (1-8) because earlier fixes can resolve later failures (e.g., `gofmt`
+fixes may resolve `lint` issues, `build` fixes resolve `test` failures).
+
+{{ references/fix-strategies.md }}
+
+### Step 5 — Re-run Preflight
```
Comment (wbreza) [inline .github/skills/azd-preflight/references/workflow.md:55]: **[:green_circle: Low]** Nested `{{ }}` include — verify fix-strategies.md is accessible

`fix-strategies.md` is included here via `{{ references/fix-strategies.md }}`, but it's not included directly from `SKILL.md`. If the framework doesn't recursively expand nested includes, the fix strategies content — essential for Step 4 — won't be inlined into the agent's context.

Verify nested includes work in the waza/Copilot framework, or add `{{ references/fix-strategies.md }}` directly to `SKILL.md` for safety.

Comment (jongio) [reply .github/skills/azd-preflight/references/workflow.md:55]: I tested this locally and it works for me.
